Update ChevronousText, Add GoLink and NewTabLink

lshig opened the pull request:
### Related issues

This PR closes https://github.com/mapbox/mr-ui/issues/13 and closes https://github.com/mapbox/mr-ui/issues/52 as adding `GoLink` (79d30c3) is dependent on adding `NewTabLink` (9059937) and updating `ChevronousText` (afb1f7a and 5103c7d).

### Description of changes

👋 Open to discussion/feedback!

**ChevronousText**
- Remove `themeIcon` for 2 reasons: it broke from the last update to `Icon` and based on how Mapbox design has traditionally used this component. The text and chevron always match in color and the text can vary in `txt-s`, `txt-m`, and `txt-l` in size, but the chevron stays the same size.
- Add `iconBefore` prop. At Mapbox we've only used `ChevronousText` for links so it's beneficial for `GoLink` to wrap the styling behaviors that `ChevronousText` offers.
- Possible future feature for this component? Allow the user to choose a `left`, `right`, `up`, or `down` chevron. Based on the current name of the component, it's not intuitive that they're only allowed to use a `left` or `right` chevron.

**NewTabLink**
- New component, examples, and tests that opinionatedly applies `rel="noopener"` to encourage more browser content-sharing security.

**GoLink**
- New component, examples, and tests that uses `ChevronousText` and `NewTabLink`. Allows opinionated variation in the following: 
  - text size (`small`, `medium`, `large`)
  - link color (`light`, `dark`)
  - bolded or unbolded text
  - whether this is a go back link (`chevron-left` before text) or a go forward link (`chevron-right` after text)
  - whether link should go directly to the `href` or open it in new tab

### Testing

- Created/updated documentation examples for each component. See the documentation Batfish site with `npm run start-docs`. Noting here that the chevron looks off in the live reload view, but otherwise normal in the built site version with `npm run build-docs` and `npm run serve-static-docs`.
- Created/updated Jest test cases and snapshots for each component. See `npm test` and the test cases app with `npm start`.

![ChevronousText test cases](https://user-images.githubusercontent.com/9087698/48283198-a85bd380-e410-11e8-9433-3f251bc6333c.png)
___
![NewTabLink test cases](https://user-images.githubusercontent.com/9087698/48283190-a2fe8900-e410-11e8-85a2-fe4bf9d2ee23.png)
___
![GoLink test cases](https://user-images.githubusercontent.com/9087698/48283204-ac87f100-e410-11e8-8683-969673cf2bec.png)

### Notes for the future

This is 1 of the series of new features (https://github.com/mapbox/mr-ui/issues/51, https://github.com/mapbox/mr-ui/issues/53, and https://github.com/mapbox/mr-ui/issues/50) I'd like to add before updating the `CHANGELOG.md` and releasing a new version of `mr-ui`.

Expect to add the following to the `CHANGELOG.md`
```
- [feature] Add **GoLink** component.
- [feature] Add **NewTabLink** component.
- **ChevronousText**
  - 🚨 [breaking change] Remove `themeIcon` prop.
  - [feature] Add `beforeIcon` prop.
```

davidtheclark commented on src/components/go-link/go-link.js:
```diff
@@ -0,0 +1,57 @@
+        <ChevronousText text={text} iconBefore={goBack} />
+      </a>
+    );
+  }
```
Hm ... it's unclear to me what value this component provides.

- The `isNewTab` & `href` props might not be necessary because the user could use a wrapper `<NewTabLink>` or `<a>` directly.
- The `color`, `size`, and `isBold` props could be removed and the user could set those props on a wrapper element.

If ChevronousText already solves the two small problems for the component (positioning the chevron and handling multiline text), and your fix in this PR allows it to be used for a link to go back — do we need this additional component?

lshig replied:
This component is limited in styling options and has a default to fall back on if links need to be added quickly. For Accounts related web apps, we either re-implement this component or copy/paste the same few lines of code as needed throughout the apps. Unfortunately that makes it very easy for unintended variation in design. It'd be ideal for us to consolidate and share the same design standards for all of our links. If someone wants more flexibility than what is standardized here, they should probably continue to create their own components or create a wrapper components around `ChevronousText`. Alternatively the Accounts team is open to adding this more opinionated/limited component to our own shared components repo similar to the Documentation team.

davidtheclark commented on src/components/new-tab-link/new-tab-link.js:
```diff
@@ -0,0 +1,31 @@
+    /** The content of the new tab link. */
+    children: PropTypes.node.isRequired,
+    /** The new tab location path. */
+    href: PropTypes.string
```
Should this prop be required? The link would have no purpose without it, right?

lshig replied:
:+1: You're right. `href` should be required and shouldn't be `""` by default. In that scenario, clicking on the link would open a new tab with the same location as where it originated from (like how Duplicate works). If a user wants that functionality, they should explicitly add it themselves. Added this change in 7a02173.

lshig commented on src/components/chevronous-text/chevronous-text.js:
```diff
@@ -4,27 +4,36 @@ import Icon from '../icon';
+        </span>{' '}
+        {textWithoutIconWord}
+      </span>
+    ) : (
```
:+1: Added this change in ddce256. Did the same for `GoLink` in 8e810e4.

lshig commented:
> Did you intend to keep this behavior (it looks like it's still working this way), or did you consider this behavior a bug? (I might lean towards considering it a bug, since it doesn't seem natural that the chevron would be the same size for very small and very large text.)

Originally this was a minor breaking change from `Icon` not having `iconTheme`. And then... I kept the behavior because there wasn't any Mapbox use cases of the text being so large that the chevron looked extra small (besides the `ChrevonousText` example in the documentation). Design-wise, it makes sense to at least allow the option to adjust the icon size in comparison to the text. I added this feature in 8e4d9f4 by adding an `iconSize` prop in `ChevronousText`. It will continue to allow a wrapper component around `ChevronousText` to determine if the icon size grows or stays at `18` by default. `Icon`'s `size` prop is also set to `18` by default.

> Do you know why the instance of `ChevronousText` in the documentation look misaligned?

The built version `lineHeight` from `setHeight()` in `Icon` returns back `24px`, while the development live server version returns back `normal`. Therefore... it won't adjust the height of the inlined `Icon`. I updated the example to ensure that a line-height was provided in 8e4d9f4, similar to `Icon`'s '`icon-b.js` example.

davidtheclark commented on src/components/chevronous-text/chevronous-text.js:
```diff
@@ -4,27 +4,46 @@ import Icon from '../icon';
+
+  static defaultProps = {
+    iconBefore: false,
+    iconSize: 18
```
I think this prop is a good solution 👍 I wonder, though, if the default value should be `1.5em` or `1.6em` to approximately match line-height by default, instead of being a hard-coded `18`? (I got those `em` ratios by reviewing the Assembly line-height/font-size ratios for the `txt-{size}` classes.) I'd imagine that in the default cause we'd want a smaller chevron if we had smaller text?

Would that also alleviate the need for special icon-size handling in GoLink?

davidtheclark replied:
Fantastic 🎿 🐻
